runtime/codegen/automarshal.go

```diff
@@ -73,12 +76,25 @@ func RegisterSerializable[T any, PT autoMarshalPointer[T]]() {
 // The returned key is stable across processes.
 func typeKey(value any) string {
 	t := reflect.TypeOf(value)
-	return t.PkgPath() + "." + t.Name()
+
+	// Embed the package path into the result since the short package name
+	// may not be unique. Note: if the registered type is a pointer, it
+	// will have an empty PkgPath, so we use the package from the pointed
+	// to type.
+	pkg := t.PkgPath()
+	if pkg == "" && t.Kind() == reflect.Pointer {
+		pkg = t.Elem().PkgPath()
+	}
+
+	return fmt.Sprintf("%s(%s)", t.String(), pkg)
 }
 
 // canEncodeInterface returns true if the concrete type of value is registered
 // as a serializable type and can therefore be sent using Encoder.Interface.
 func canEncodeInterface(value any) bool {
+	if _, ok := value.(AutoMarshal); !ok {
```

Comment:
If I'm understanding correctly, if `value` is of type `T`, `T` implements `error`, but `*T` implements `AutoMarshal` instead of `T`, then we return false here? Is it possible for us to encode `value` by taking a pointer to it? 

I noticed that values of type `codegen.customTestError` return false here and end up being decoded as `codegen.decodedError`, but I would have expected them be decoded into a `codegen.customTestError`.

Reply:
Good catch. Fixed.